frontend/controller/actions/group.js

```diff
@@ -283,21 +257,41 @@ export default (sbp('sbp/selectors/register', {
   // secret keys to be shared with us, (b) ready to call the inviteAccept
   // action if we haven't done so yet (because we were previously waiting for
   // the keys), or (c) already a member and ready to interact with the group.
-  'gi.actions/group/join': async function (params: $Exact<ChelKeyRequestParams> & { blockOriginatingContract?: boolean }) {
+  'gi.actions/group/join': async function (params: $Exact<ChelKeyRequestParams>) {
+    // We want to process any current events first, before setting
+    // JOINING_GROUP, so that we process leave actions and don't interfere
+    // with the leaving process (otherwise, the side-effects will prevent
+    // us from fully leaving).
+    await sbp('chelonia/contract/wait', params.contractID)
     sbp('okTurtles.data/set', 'JOINING_GROUP-' + params.contractID, true)
     try {
-      const rootState = sbp('state/vuex/state')
-      const username = rootState.loggedIn.username
-      const userID = rootState.loggedIn.identityContractID
+      const { loggedIn } = sbp('state/vuex/state')
+      if (!loggedIn) throw new Error('[gi.actions/group/join] Not logged in')
 
-      console.log('@@@@@@@@ AT join for ' + params.contractID)
+      const { username, identityContractID: userID } = loggedIn
 
       await sbp('chelonia/contract/sync', params.contractID)
+      const rootState = sbp('state/vuex/state')
+      if (!rootState.contracts[params.contractID]) {
+        console.warn('[gi.actions/group/join] The group contract was removed after sync. If this happened during logging in, this likely means that we left the group on a different session.', { contractID: params.contractID })
+        return
+      }
 
-      if (rootState.contracts[params.contractID]?.type !== 'gi.contracts/group') {
+      if (rootState.contracts[params.contractID].type !== 'gi.contracts/group') {
         throw Error(`Contract ${params.contractID} is not a group`)
       }
 
+      // At this point, we do not know whether we should continue with the
+      // join process, because we don't know where we stand in the process.
+      // One edge case we need to handle is that according to our records
+      // (in the identity contract) we're members, and we have completed
+      // all of the steps for joining in the past, but we've then been removed
+      // while we were offline. In this case, we should *not* re-join
+      // automatically, even if we have a valid invitation secret and are
+      // technically able to. However, if the previous situation we *should*
+      // attempt to rejoin if the action was user-initiated.
```

Comment:
Think there's a typo here:

> However, if the previous situation we *should* attempt to rejoin if the action was user-initiated.

Instead of "if the previous situation", do you mean "in the previous situation"? Or "if in the previous situation"? Maybe you can just rewrite that sentence?